File: laravel/app/Models/Experience.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Casts\Attribute;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsToMany;
use Illuminate\Database\Eloquent\Relations\HasMany;

class Experience extends Model
{
    protected $fillable = [
        'company',
        'title',
        'start_date',
        'end_date',
        'summary',
    ];

    protected $casts = [
        'start_date' => 'date',
        'end_date' => 'date',
    ];

    protected $appends = ['is_current', 'duration_months'];

    public function skills(): BelongsToMany
    {
        return $this->belongsToMany(Skill::class, 'experience_skills');
    }

    public function accomplishments(): HasMany
    {
        return $this->hasMany(ExperienceAccomplishment::class)->orderBy('sort_order');
    }

    protected function isCurrent(): Attribute
    {
        return Attribute::get(fn () => $this->end_date === null);
    }

    protected function durationMonths(): Attribute
    {
        return Attribute::get(
            fn () => $this->start_date->diffInMonths($this->end_date ?? now())
        );
    }
}

File: rails-admin/app/models/experience.rb
# frozen_string_literal: true

class Experience < ApplicationRecord
  has_many :accomplishments,
           class_name: 'ExperienceAccomplishment',
           dependent: :destroy,
           inverse_of: :experience

  has_and_belongs_to_many :skills, join_table: 'experience_skills'

  validates :company, :title, :start_date, presence: true
  validate :end_date_after_start_date

  scope :current, -> { where(end_date: nil) }
  scope :past, -> { where.not(end_date: nil) }
  scope :chronological, -> { order(start_date: :desc) }

  # is_current is a generated column in PostgreSQL
  def current?
    end_date.nil?
  end

  def duration_months
    end_date_or_now = end_date || Date.current
    ((end_date_or_now.year - start_date.year) * 12) +
      (end_date_or_now.month - start_date.month)
  end

  def duration_display
    months = duration_months
    years = months / 12
    remaining = months % 12

    parts = []
    parts << "#{years}y" if years.positive?
    parts << "#{remaining}m" if remaining.positive?
    parts.join(' ') || '< 1m'
  end

  def display_name
    "#{title} at #{company}"
  end

  private

  def end_date_after_start_date
    return unless end_date && start_date
    return unless end_date < start_date

    errors.add(:end_date, 'must be after start date')
  end
end
